softlayer/resource_softlayer_bare_metal.go

```diff
@@ -0,0 +1,442 @@
+	if len(ssh_key_ids) > 0 {
+		hardware.SshKeys = make([]datatypes.Security_Ssh_Key, 0, len(ssh_key_ids))
+		for _, ssh_key_id := range ssh_key_ids {
+			hardware.SshKeys = append(hardware.SshKeys, datatypes.Security_Ssh_Key{
+				Id: sl.Int(ssh_key_id.(int)),
+			})
+		}
+	}
+
+	return hardware, nil
+}
+
+func resourceSoftLayerBareMetalCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(*session.Session)
+	hwService := services.GetHardwareService(sess)
+	orderService := services.GetProductOrderService(sess)
+
+	hardware, err := getBareMetalOrderFromResourceData(d, meta)
+	if err != nil {
+		return err
+	}
+
+	order, err := hwService.GenerateOrderTemplate(&hardware)
+	if err != nil {
+		return fmt.Errorf(
+			"Encountered problem trying to get the bare metal order template: %s", err)
+	}
+
+	// Set image template id if it exists
+	if rawImageTemplateId, ok := d.GetOk("image_template_id"); ok {
+		imageTemplateId := rawImageTemplateId.(int)
+		order.ImageTemplateId = sl.Int(imageTemplateId)
+	}
+
+	log.Println("[INFO] Ordering bare metal server")
+
+	_, err = orderService.PlaceOrder(&order, sl.Bool(false))
+	if err != nil {
+		return fmt.Errorf("Error ordering bare metal server: %s", err)
+	}
+
+	log.Printf("[INFO] Bare Metal Server ID: %s", d.Id())
+
+	// wait for machine availability
+	bm, err := WaitForBareMetalProvision(&hardware, meta)
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for bare metal server (%s) to become ready: %s", d.Id(), err)
+	}
+
+	d.SetId(fmt.Sprintf("%d", *bm.(datatypes.Hardware).Id))
+
+	return resourceSoftLayerBareMetalRead(d, meta)
+}
+
+func resourceSoftLayerBareMetalRead(d *schema.ResourceData, meta interface{}) error {
+	service := services.GetHardwareService(meta.(*session.Session))
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid ID, must be an integer: %s", err)
+	}
+
+	result, err := service.Id(id).Mask(
+		"hostname,domain," +
+			"primaryIpAddress,primaryBackendIpAddress,privateNetworkOnlyFlag," +
+			"userData[value]," +
+			"hourlyBillingFlag," +
+			"datacenter[id,name,longName]," +
+			"primaryNetworkComponent[networkVlan[id,primaryRouter,vlanNumber],maxSpeed]," +
+			"primaryBackendNetworkComponent[networkVlan[id,primaryRouter,vlanNumber],maxSpeed]",
+	).GetObject()
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving bare metal server: %s", err)
+	}
+
+	d.Set("hostname", *result.Hostname)
+	d.Set("domain", *result.Domain)
+
+	if result.Datacenter != nil {
+		d.Set("datacenter", *result.Datacenter.Name)
+	}
+
+	d.Set("network_speed", *result.PrimaryNetworkComponent.MaxSpeed)
+	if result.PrimaryIpAddress != nil {
+		d.Set("public_ipv4_address", *result.PrimaryIpAddress)
+	}
+	d.Set("private_ipv4_address", *result.PrimaryBackendIpAddress)
+
+	d.Set("private_network_only", *result.PrivateNetworkOnlyFlag)
+	d.Set("hourly_billing", *result.HourlyBillingFlag)
+
+	if result.PrimaryNetworkComponent.NetworkVlan != nil {
```

Comment:
Have you confirmed that `PrimaryNetworkComponent` is non-nil for a private-only BM, or should we guard against nil for this field as well?